crypto/src/proof_of_work.rs

```diff
@@ -0,0 +1,152 @@
+use num_bigint::BigUint;
+use hex::{FromHex, FromHexError};
+use std::convert::TryFrom;
+use sodiumoxide::randombytes::randombytes;
+use super::{blake2b, crypto_box::PublicKey};
+
+#[derive(Clone, PartialEq)]
+pub struct ProofOfWork([u8; 0x18]);
+
+impl AsRef<[u8]> for ProofOfWork {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl FromHex for ProofOfWork {
+    type Error = FromHexError;
+
+    fn from_hex<T: AsRef<[u8]>>(hex: T) -> Result<Self, Self::Error> {
+        let bytes = hex::decode(hex)?;
+        let mut arr = [0u8; 0x18];
+        arr.copy_from_slice(&bytes);
+        Ok(ProofOfWork(arr))
+    }
+}
+
+impl ProofOfWork {
+    pub const DEFAULT_TARGET: f64 = 24.0;
+
+    pub fn generate(public_key: &PublicKey, target: f64) -> Self {
+        let mut data = [0; 0x20 + 0x18];
+        data[..0x20].clone_from_slice(public_key.as_ref().as_ref());
+        data[0x20..].clone_from_slice(randombytes(0x18).as_ref());
+
+        let target_number = make_target(target);
+        loop {
+            if let Ok(()) = check_proof_of_work_inner(data.as_ref(), &target_number) {
+                let mut nonce = [0; 0x18];
+                nonce.clone_from_slice(&data[0x20..]);
+                return ProofOfWork(nonce);
+            } else {
+                // the code might look obscure,
+                // but it just treat `data[0x20..0x38]` as an 192-bit integer and increment it
+
+                let mut c = u64::from_be_bytes(<[u8; 8]>::try_from(&data[0x30..0x38]).unwrap());
+                if c == u64::MAX {
+                    let mut b = u64::from_be_bytes(<[u8; 8]>::try_from(&data[0x28..0x30]).unwrap());
+                    if b == u64::MAX {
+                        let mut a = u64::from_be_bytes(<[u8; 8]>::try_from(&data[0x20..0x28]).unwrap());
+                        if a == u64::MAX {
+                            a = 0;
+                            b = 0;
+                            c = 0;
+                        } else {
+                            a += 1;
+                            b = 0;
+                            c = 0;
+                        }
+                        data[0x20..0x28].clone_from_slice(a.to_be_bytes().as_ref());
+                    } else {
+                        b += 1;
+                        c = 0;
+                    }
+                    data[0x28..0x30].clone_from_slice(b.to_be_bytes().as_ref());
+                } else {
+                    c += 1;
+                }
+                data[0x30..0x38].clone_from_slice(c.to_be_bytes().as_ref());
+            }
+        }
+    }
+
+    pub fn check(&self, pk: &[u8; 0x20], target: f64) -> Result<(), ()> {
```

Comment:
@vlad9486 
do you use this somewhere?
if not, please remove, or remove just pub,

if anybody wants to use it, he would wonder about why `&[u8; 0x20]`, because probably he will have just some PubKey or Vec<u8>